### PR TITLE
Implement dependency report

### PR DESCRIPTION
**Pull request by mmanciop**
If the Lumigo OpenTelemetry distro for Python is configured to use the default Lumigo endpoint and has the LUMIGO_TRACER_TOKEN env var set, report the Node.js dependencies we can discover to the Lumigo backend for analysis using the <DEFAULT_ENDPOINT>/v1/dependencies endpoint to be provided by the Lumigo SaaS.

### CODE REVIEW

File: src/test/unit/dependencies/test_dependencies.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        assert url == "https://ga-otlp.lumigo-tracer-edge.golumigo.com/v1/dependencies"
+        assert lumigo_token == "abcdef"
+
+        parsed_data = loads(data)
+        resource_attributes = parsed_data["resourceAttributes"]
+        dependencies = parsed_data["dependencies"]
+
+        assert resource_attributes and resource_attributes["lumigo.distro.version"]
+        assert "process.environ" not in resource_attributes.keys()
+        assert dependencies and len(dependencies) > 0
```
Comment (nirLumigo) [inline src/test/unit/dependencies/test_dependencies.py:70]: Great test cases, many cases are covered 🙂 
Can you also add a test that verifies that an error from the endpoint doesn't cause an exception?

**Review by nirLumigo — CHANGES_REQUESTED** [review]
Nice! :)

File: src/lumigo_opentelemetry/dependencies/__init__.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    data = dumps(
+        {
+            "resourceAttributes": _prepare_resource_attributes_for_marshalling(
+                resource_attributes
+            ),
+            "dependencies": dependencies,
+        }
+    )
+
+    _report_to_saas(url, lumigo_token, data)
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/lumigo_opentelemetry/dependencies/__init__.py:32]: Not worth it for now. The startup of ECS containers is anyhow something you measure in tens of seconds, and until we have a good reason to mess with it (i.e. end users complaints), going down the `multiprocessing` route and the bundle of joy that is not knowing is it is compiled into CPython is something I would not deal with.

File: src/lumigo_opentelemetry/__init__.py [hunk 3]
```diff
@@ -114,6 +129,24 @@ def init() -> Dict[str, Any]:
                 ),
             )
         )
+
+        if (
+            lumigo_report_dependencies == "true"
+            and lumigo_endpoint == DEFAULT_LUMIGO_ENDPOINT
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/lumigo_opentelemetry/__init__.py:135]: No, specifically there must not be. See the guarantees we provide in the README.md.

File: src/lumigo_opentelemetry/__init__.py [hunk 4]
```diff
@@ -114,6 +129,24 @@ def init() -> Dict[str, Any]:
...
+        ):
+            from lumigo_opentelemetry.dependencies import report
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/lumigo_opentelemetry/__init__.py:137]: There is no real reason to do that until we know we are going to use it. This way, if dependency reporting is turned off, the interpreter does less work with most settings.

File: src/lumigo_opentelemetry/__init__.py [hunk 5]
```diff
@@ -114,6 +129,24 @@ def init() -> Dict[str, Any]:
...
+
+            try:
+                # TODO Avoid sending the process env and non-infrastructure resource attributes
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/lumigo_opentelemetry/__init__.py:140]: Actually I did fix the issue and forgot the comment

File: src/lumigo_opentelemetry/__init__.py [hunk 6]
```diff
@@ -114,6 +129,24 @@ def init() -> Dict[str, Any]:
...
+                report(
+                    DEFAULT_DEPENDENCIES_ENDPOINT,
+                    lumigo_token,
+                    infrastructure_resource.attributes,
+                )
+            except Exception as e:
+                logger.debug("Cannot report dependencies to Lumigo", e)
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/lumigo_opentelemetry/__init__.py:147]: No. Reporting dependencies is a best-effort thing. We actively do not want to freak out the user when it fails.

File: src/test/unit/dependencies/test_dependencies.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
+from unittest import TestCase
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/test/unit/dependencies/test_dependencies.py:1]: I can go either way here. Done in 21be272f7a3e613ed003ab62f791a7b185115e8d.

File: src/test/unit/dependencies/test_dependencies.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+from unittest.mock import patch
+
+from json import loads
+from os import environ
+
+from lumigo_opentelemetry import init
+
+
+class TestDependencyReport(TestCase):
+    @patch("lumigo_opentelemetry.dependencies.report")
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/test/unit/dependencies/test_dependencies.py:11]: You mean `requests_mock`, right?

[the remark this answers is not in the document]
Comment (mmanciop) [reply src/test/unit/dependencies/test_dependencies.py]: Turns out https://httpretty.readthedocs.io/en/latest/api.html#last-request is every bit as good as `requests-mock` and works with `http.client`. Done in 9dbc843.

File: src/test/unit/dependencies/test_dependencies.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/test/unit/dependencies/test_dependencies.py]: Done in 1c96abe.

Comment (mmanciop) [reply src/test/unit/dependencies/test_dependencies.py]: Done in fb3d67c. I'll see if I can actually throw in one for the timeout too.

File: src/lumigo_opentelemetry/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nirLumigo) [reply src/lumigo_opentelemetry/__init__.py:135]: We can still keep the guarantees, if by default we won't send to another endpoint but we will allow to send it to another endpoint by adding an env var, something like `LUMIGO_ALLOW_CUSTOM_DEPENDENCIES=true`.

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -8,26 +8,36 @@ repos:
...
 -   repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v0.971
+    rev: v0.982
     hooks:
-    -   id: mypy
-        entry: venv/bin/mypy
-        exclude: '(src/test/|noxfile\.py)|(src/lumigo_opentelemetry/external/.*)'
-        args: ['--explicit-package-bases',  # Needed for src/ci
-               '--namespace-packages',  # Needed for opentelemetry package
-               '--disallow-any-generics', '--disallow-untyped-defs', '--disallow-incomplete-defs', '--check-untyped-defs', '--no-implicit-optional', '--warn-redundant-casts', '--warn-unused-ignores', '--warn-return-any', '--no-implicit-reexport', '--strict-equality', '--ignore-missing-imports']
```
Comment (nirLumigo) [inline .pre-commit-config.yaml:18]: Why did you remove "warn-unused-ignores"?

File: src/lumigo_opentelemetry/dependencies/__init__.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+) -> Dict[str, Any]:
+    return {
+        attribute_name: resource_attributes[attribute_name]
+        for attribute_name in resource_attributes.keys()
+    }
+
+
+def _report_to_saas(url: str, lumigo_token: str, data: str):
+    parsed_url = urlparse(url)
+    timeout = environ.get("LUMIGO_CONNECTION_TIMEOUT", DEFAULT_CONNECTION_TIMEOUT)
```
[the remark this answers is not in the document]
Comment (nirLumigo) [reply src/lumigo_opentelemetry/dependencies/__init__.py:55]: ?

File: src/lumigo_opentelemetry/dependencies/__init__.py [hunk 11]
```diff
@@ -51,28 +50,23 @@ def _prepare_resource_attributes_for_marshalling(
...
+    response_status = response.status
 
-        if response_status != 200:
-            logger.debug(
-                "Dependency report failed with status code %s", response_status
-            )
-    except Exception as e:
-        logger.debug("Dependency report failed", e)
+    if response_status != 200:
+        raise Exception(f"Dependency report failed with status code {response_status}")
```
Comment (nirLumigo) [inline src/lumigo_opentelemetry/dependencies/__init__.py:72]: Can you add here more data? Maybe something like `response.text`.
In case of errors, it will make debugging easier.

File: src/test/unit/test_tracer.py [hunk 12]
```diff
@@ -9,3 +19,83 @@ def test_access_trace_provider(self):
...
+    @httpretty.activate(allow_net_connect=False)
+    @patch.dict(
+        environ, {"LUMIGO_TRACER_TOKEN": "", "LUMIGO_REPORT_DEPENDENCIES": "true"}
+    )
+    def test_dependency_report_disabled_if_no_lumigo_token(self):
+        assert not environ.get("LUMIGO_TRACER_TOKEN")
+
+        init()
+
+        assert not httpretty.latest_requests()
```
Comment (nirLumigo) [inline src/test/unit/test_tracer.py:34]: 😍

File: src/test/unit/test_tracer.py [hunk 13]
```diff
@@ -9,3 +19,83 @@ def test_access_trace_provider(self):
...
+            "https://ga-otlp.lumigo-tracer-edge.golumigo.com/v1/dependencies",
+            headers={"Authentication": "LumigoToken abcdef"},
+            status_code=200,
+        )
+
+        init()
+
+        # TODO because of the autoload mechanics, that assume that init()
+        # is only called implicitly on import, here technically we execute
+        # init twice, and we check only the latest request.
```
Comment (nirLumigo) [inline src/test/unit/test_tracer.py:93]: Why is `init` executed twice? It is being called only once in the test.

**Review by nirLumigo — CHANGES_REQUESTED** [review]
Great! :)

File: src/lumigo_opentelemetry/dependencies/__init__.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    data = dumps(
+        {
+            "resourceAttributes": _prepare_resource_attributes_for_marshalling(
+                resource_attributes
+            ),
+            "dependencies": dependencies,
+        }
+    )
+
+    _report_to_saas(url, lumigo_token, data)
```
Comment (nirLumigo) [inline src/lumigo_opentelemetry/dependencies/__init__.py:41]: Can you add a log message saying that the dependencies were reported?

File: .pre-commit-config.yaml (hunk 9 above)
Comment (mmanciop) [reply .pre-commit-config.yaml:18]: Because of a false positive I could not get rid of otherwise.

File: src/lumigo_opentelemetry/dependencies/__init__.py (hunk 14 above)
Comment (mmanciop) [reply src/lumigo_opentelemetry/dependencies/__init__.py:41]: No, it's noise in debugging

File: src/lumigo_opentelemetry/dependencies/__init__.py (hunk 11 above)
Comment (mmanciop) [reply src/lumigo_opentelemetry/dependencies/__init__.py:72]: I think it is because of https://github.com/lumigo-io/opentelemetry-python-distro/blob/6f310a6e1357896c86c2ba253bcdb4659ca705d1/src/lumigo_opentelemetry/__init__.py#L169, which we need for `autowrapt` to work

File: src/test/unit/test_tracer.py (hunk 13 above)
Comment (mmanciop) [reply src/test/unit/test_tracer.py:93]: I think it is because of

[opentelemetry-python-distro/src/lumigo_opentelemetry/__init__.py](https://github.com/lumigo-io/opentelemetry-python-distro/blob/6f310a6e1357896c86c2ba253bcdb4659ca705d1/src/lumigo_opentelemetry/__init__.py#L169)

Line 169 in [6f310a6](https://github.com/lumigo-io/opentelemetry-python-distro/commit/6f310a6e1357896c86c2ba253bcdb4659ca705d1)
 init_data = init() 
, which we need for autowrapt to work

File: src/lumigo_opentelemetry/dependencies/__init__.py (hunk 11 above)
Comment (mmanciop) [reply src/lumigo_opentelemetry/dependencies/__init__.py]: Done in b6b3e2d

File: src/lumigo_opentelemetry/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/lumigo_opentelemetry/__init__.py:135]: We decided not to do this, and instead using the `--add-host` option of Docker in our tests.

File: .pre-commit-config.yaml (hunk 9 above)
Comment (nirLumigo) [reply .pre-commit-config.yaml:18]: It means that you can remove the `# type: ignore`.

File: src/lumigo_opentelemetry/dependencies/__init__.py (hunk 14 above)
Comment (nirLumigo) [reply src/lumigo_opentelemetry/dependencies/__init__.py:41]: Tracers are much harder to debug than services in the backend, therefore we try to add much more log messages than usual, it proved to be useful in the past.
